### PR TITLE
[JENKINS-33944] User-scoped credentials cannot be looked up

### PR DESCRIPTION
**Pull request by armfergom**
See description of the issue in [JENKINS-33944](https://issues.jenkins-ci.org/browse/JENKINS-33944).

Work included:

- [x] Fix issue.
- [x] Implement test.
- [x] Migrate test suite to new JTH.
- [x] Add extra test cases (covering cases not related to JENKINS-33944)

@reviewbybees esp. @stephenc

### CODE REVIEW

File: src/main/java/com/cloudbees/plugins/credentials/UserCredentialsProvider.java [hunk 1]
```diff
@@ -145,8 +145,8 @@
                 UserCredentialsProperty property = user.getProperty(UserCredentialsProperty.class);
                 if (property != null) {
                     // we need to impersonate if the requesting authentication is not the current authentication.
-                    boolean needImpersonation = user.equals(User.current());
-                    SecurityContext old = needImpersonation ? null : ACL.impersonate(user.impersonate());
+                    boolean needImpersonation = !user.equals(User.current());
+                    SecurityContext old = needImpersonation ? ACL.impersonate(user.impersonate()) : null;
```
Comment (jglick) [inline src/main/java/com/cloudbees/plugins/credentials/UserCredentialsProvider.java]: Maybe I am being blinded, but is this just a no-op, equivalent to renaming the variable?

Comment (jglick) [inline src/main/java/com/cloudbees/plugins/credentials/UserCredentialsProvider.java]: Never mind, used below in the `finally` block.

**Comment by jglick** [comment]
From 3876043c2051fb1afdc9abab89bc6653b7f74ad3 I guess. Review by @stephenc is mandatory.

**Comment by armfergom** [comment]
@jglick correct, that's the commit that introduced what, IMHO, is a regression. Forgot to mention @stephenc for the review. Thanks!

**Comment by stephenc** [comment]
@armfergom can you update the wiki page changelog with details of the 1.27 release

**Comment by armfergom** [comment]
@stephenc sure thing!

**Pull request merged**